### PR TITLE
Add OAEP KeyWrap support

### PR DESCRIPTION
**Pull request by cipherboy**
This is a patch set to add OAEP support for KeyWrap and KeyUnwrap. This will need to wait to merge until NSS supports OAEP at the high-level `PK11_*` function call level (for key wrapping). A patchset for NSS is available here to test against: [`cipherboy/nss@add-oaep-support`](https://github.com/cipherboy/nss/pull/5). 

TODO:

 - [x] Propose NSS patchset upstream so we can merge this PR: https://bugzilla.mozilla.org/show_bug.cgi?id=1666891
 - [x] Add PKCS#11 mech
 - [x] Add parameters similar to KBKDF
 - [x] Add to KeyWrap
 - [x] Minor improvements to build system. :-) 
 - [x] Update to fallback to original code when OAEP isn't present. 

Adding `Cipher` support for OAEP will be done at a later time in a separate PR.

### CODE REVIEW

**Comment by cipherboy** [comment]
(Build failures are expected since this requires updates to NSS first).

**Comment by cipherboy** [comment]
@jmagne This has been updated so that when OAEP isn't present, we'll fallback to the old code -- and throw an error on OAEP operations.

File: org/mozilla/jss/pkcs11/PK11KeyWrapper.c [hunk 1]
```diff
@@ -17,6 +17,8 @@
 #include <jssutil.h>
 #include <pk11util.h>
 #include <Algorithm.h>
+#include "StaticVoidPointer.h"
+#include "jssconfig.h"
```
Comment (cipherboy) [inline org/mozilla/jss/pkcs11/PK11KeyWrapper.c:21]: @jmagne -- sorry, I forgot one line, that's why it was failing. :-) 

`jssconfig.h` defines `HAVE_NSS_OAEP` from the build system. Now it works again.

File: org/mozilla/jss/crypto/JSSOAEPParameterSpec.c [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+    jobject params_obj;
+
+    this_class = (*env)->GetObjectClass(env, this);
+    if (this_class == NULL) {
+        return;
+    }
+
+    if (oaep_GetHashAlg(env, this, this_class, &hashAlg) != PR_SUCCESS) {
+        goto failure;
```
Comment (edewata) [inline org/mozilla/jss/crypto/JSSOAEPParameterSpec.c:90]: I think we can `return` immediately here, but no big deal.

File: org/mozilla/jss/crypto/JSSOAEPParameterSpec.c [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    if (oaep_GetMGFType(env, this, this_class, &mgf) != PR_SUCCESS) {
+        goto failure;
```
Comment (edewata) [inline org/mozilla/jss/crypto/JSSOAEPParameterSpec.c:94]: Same thing here.

File: org/mozilla/jss/crypto/JSSOAEPParameterSpec.java [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+ *
+ * Note that this class cannot be used with other JCA providers natively.
+ */
+public class JSSOAEPParameterSpec extends NativeEnclosure implements AlgorithmParameterSpec {
+    public long hashAlg;
+    public long mgf;
+    public byte[] sourceData;
+
+    public JSSOAEPParameterSpec(String mdName, String mgfName, AlgorithmParameterSpec mgfSpec, PSource pSrc) {
+        super();
```
Comment (edewata) [inline org/mozilla/jss/crypto/JSSOAEPParameterSpec.java:28]: The `super()` is redundant. It's called implicitly. No big deal though.

File: org/mozilla/jss/crypto/JSSOAEPParameterSpec.java [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+
+        setDigestAlgorithm(mdName);
+        setMaskGenAlgorithm(mgfName);
+        setMaskGenAlgorithmType(mgfSpec);
+        setPSource(pSrc);
+    }
+
+    public JSSOAEPParameterSpec(OAEPParameterSpec copy) {
+        super();
```
Comment (edewata) [inline org/mozilla/jss/crypto/JSSOAEPParameterSpec.java:37]: Same thing here.

File: org/mozilla/jss/crypto/JSSOAEPParameterSpec.java [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+    public void setPSource(PSource spec) throws IllegalArgumentException {
+        if (spec == null) {
+            return;
+        }
+
+        if (!(spec instanceof PSource.PSpecified)) {
+            String msg = "Expected PSource spec to be an instance of ";
+            msg += "PSource.PSpecified, but wasn't: " + spec;
+            throw new IllegalArgumentException(msg);
+        }
```
Comment (edewata) [inline org/mozilla/jss/crypto/JSSOAEPParameterSpec.java:174]: Is this necessary? Java will throw `ClassCastException` automatically in the following code.

File: org/mozilla/jss/crypto/JSSOAEPParameterSpec.java [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+    }
+
+    public void setMaskGenAlgorithmType(long algo) throws IllegalArgumentException {
+        mgf = algo;
+    }
+
+    public void setPSource(PSource spec) throws IllegalArgumentException {
+        if (spec == null) {
+            return;
```
Comment (edewata) [inline org/mozilla/jss/crypto/JSSOAEPParameterSpec.java:167]: Do we want to set `sourceData = null` here? No big deal, just semantics.

**Review by edewata — COMMENTED** [review]
I just have a few minor comments. I'd defer to @jmagne or @ladycfu for the ACK.

File: org/mozilla/jss/crypto/JSSOAEPParameterSpec.c (hunk 3 above)
Comment (cipherboy) [reply org/mozilla/jss/crypto/JSSOAEPParameterSpec.c:94]: ACK, I've added a comment after this one mentioning allocated data.

File: org/mozilla/jss/crypto/JSSOAEPParameterSpec.java (hunk 6 above)
Comment (cipherboy) [reply org/mozilla/jss/crypto/JSSOAEPParameterSpec.java:174]: I think the `IllegalArgumentException` is nicer than the `ClassCastException` personally in this case.

**Review by jmagne — APPROVED** [review]
Providing a quick review based on the fact that I've been actually using this code for quite some time. So It has shown itself to work just fine in practice, while trying to integrate this stuff into pki.

The code itself looks fine to m, but I haven't gone down to the micro level to find memory mistakes and all that. I trust the various mem checking such as valgrind has been run on this, so it should be fine.  Given all that, approving...

**Comment by cipherboy** [comment]
@jmagne As far as I can tell, there's only two false-positive, both from JVM internals:

```
==707323== 24 bytes in 1 blocks are still reachable in loss record 178 of 1,083
==707323==    at 0x483A809: malloc (vg_replace_malloc.c:307)
==707323==    by 0x582F6B4: os::malloc(unsigned long, MemoryType, NativeCallStack const&) (in /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.275.b01-1.fc32.x86_64/jre/lib/amd64/server/libjvm.so)
==707323==    by 0x54F9D9E: InstanceKlass::get_jmethod_id(instanceKlassHandle, methodHandle) (in /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.275.b01-1.fc32.x86_64/jre/lib/amd64/server/libjvm.so)
==707323==    by 0x55CF326: get_method_id(JNIEnv_*, _jclass*, char const*, char const*, bool, Thread*) [clone .constprop.0] (in /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.275.b01-1.fc32.x86_64/jre/lib/amd64/server/libjvm.so)
==707323==    by 0x55CF869: jni_GetMethodID (in /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.275.b01-1.fc32.x86_64/jre/lib/amd64/server/libjvm.so)
==707323==    by 0x38F7ADCA: JSS_PR_wrapStaticVoidPointer (StaticVoidPointer.c:27)
==707323==    by 0x38F62DF9: Java_org_mozilla_jss_crypto_JSSOAEPParameterSpec_acquireNativeResources (JSSOAEPParameterSpec.c:111)
==707323==    by 0x621B426: ???
==707323==    by 0x620AFFC: ???
==707323==    by 0x620AFFC: ???
==707323==    by 0x620AE53: ???
==707323==    by 0x62034E6: ???

----- snip -----

==707323== 48 bytes in 1 blocks are still reachable in loss record 424 of 1,083
==707323==    at 0x483A809: malloc (vg_replace_malloc.c:307)
==707323==    by 0x582F6B4: os::malloc(unsigned long, MemoryType, NativeCallStack const&) (in /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.275.b01-1.fc32.x86_64/jre/lib/amd64/server/libjvm.so)
==707323==    by 0x59AE8C0: Symbol::operator new(unsigned long, int, Thread*) (in /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.275.b01-1.fc32.x86_64/jre/lib/amd64/server/libjvm.so)
==707323==    by 0x59AF93F: SymbolTable::allocate_symbol(unsigned char const*, int, bool, Thread*) (in /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.275.b01-1.fc32.x86_64/jre/lib/amd64/server/libjvm.so)
==707323==    by 0x59B04EC: SymbolTable::basic_add(int, unsigned char*, int, unsigned int, bool, Thread*) (in /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.275.b01-1.fc32.x86_64/jre/lib/amd64/server/libjvm.so)
==707323==    by 0x59B06C4: SymbolTable::lookup(char const*, int, Thread*) (in /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.275.b01-1.fc32.x86_64/jre/lib/amd64/server/libjvm.so)
==707323==    by 0x55CD24D: jni_FindClass (in /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.275.b01-1.fc32.x86_64/jre/lib/amd64/server/libjvm.so)
==707323==    by 0x38F7ADA5: JSS_PR_wrapStaticVoidPointer (StaticVoidPointer.c:21)
==707323==    by 0x38F62DF9: Java_org_mozilla_jss_crypto_JSSOAEPParameterSpec_acquireNativeResources (JSSOAEPParameterSpec.c:111)
==707323==    by 0x621B426: ???
==707323==    by 0x620AFFC: ???
==707323==    by 0x620AFFC: ???
```

Note that both `jni_GetMethodID`'s and `jni_FindClass`'s results should not be explicitly freed and should be handled by the JVM.

Other than that, there's no OAEP-specific leaks.

**Pull request merged**